test/extensions/common/dynamic_forward_proxy/dns_cache_impl_test.cc

```diff
@@ -954,10 +969,101 @@ TEST(DnsCacheConfigOptionsTest, NonEmptyDnsResolutionConfig) {
   dns_resolvers->mutable_socket_address()->set_address("1.2.3.4");
   dns_resolvers->mutable_socket_address()->set_port_value(8080);
 
-  std::vector<Network::Address::InstanceConstSharedPtr> expected_dns_resolvers;
-  expected_dns_resolvers.push_back(Network::Address::resolveProtoAddress(*dns_resolvers));
+  envoy::config::core::v3::TypedExtensionConfig typed_dns_resolver_config;
+  envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+  cares.add_resolvers()->MergeFrom(*dns_resolvers);
+  typed_dns_resolver_config.mutable_typed_config()->PackFrom(cares);
+  typed_dns_resolver_config.set_name(Envoy::Network::cares_dns_resolver);
   EXPECT_CALL(dispatcher,
-              createDnsResolver(CustomDnsResolversSizeEquals(expected_dns_resolvers), _))
+              createDnsResolver(CustomTypedDnsResolverConfigEquals(typed_dns_resolver_config)))
+      .WillOnce(Return(resolver));
+  DnsCacheImpl dns_cache_(dispatcher, tls, random, loader, store, config);
+}
+
+// Test dns_resolution_config is in place, use it and overriding use_tcp_for_dns_lookups.
+TEST(DnsCacheConfigOptionsTest, NonEmptyDnsResolutionConfigOverridingUseTcp) {
+  NiceMock<Event::MockDispatcher> dispatcher;
+  std::shared_ptr<Network::MockDnsResolver> resolver{std::make_shared<Network::MockDnsResolver>()};
+  NiceMock<ThreadLocal::MockInstance> tls;
+  NiceMock<Random::MockRandomGenerator> random;
+  NiceMock<Runtime::MockLoader> loader;
+  Stats::IsolatedStoreImpl store;
+  envoy::extensions::common::dynamic_forward_proxy::v3::DnsCacheConfig config;
+
+  // setup use_tcp
+  config.set_use_tcp_for_dns_lookups(false);
+
+  // setup dns_resolution_config
+  envoy::config::core::v3::Address* dns_resolvers =
```

Comment:
can you factor this with the similar block in the test below? I think they differ only by a true/false which could be a bool passed to a helper method in the test class.

Reply:
I went through these test code. It is true that these testing code has a lot of similarities in general. Like some testings differ only by one proto field, or some testing differ only by different value. It is possible to implement one function to test a lot of different cases with parameters and combined if ... else, or swtich ... case. But, I would prefer keep these testing functions as plain as it is, which is also the case in the original code.  Does this make sense?

Reply:
I think Josh meant to add helper function for adding the resolver.

Reply:
For helper function to add resolvers , we need to pass in the IP address, or port # et.al. 
Maybe it's not a bad idea to keep the testing function plain so it's easy to understand what is trying to test, also no need to worry about the bugs in it?

Reply:
ACK

Reply:
Replaced the resolver creating code with Network::Utility::addressToProtobufAddress(), and used this in most of the testing code.